### PR TITLE
Encoder Pose Estimation

### PR DESCRIPTION
**Pull request by 4ndr3w**
Velocity-based pose estimation

### CODE REVIEW

File: src/magellan_firmware/firmware/magellan_controller/Drivetrain.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    steering_angle_msg_.data = GetTurningRadius(percent);
+    steering_angle_publisher_.publish(&steering_angle_msg_);
+}
+
+void Drivetrain::SetSteeringAngle(double angle) {
+    SetSteeringPercent(GetPercentForSteeringAngle(angle));
+}
+
+double Drivetrain::GetSteeringAngleForPercent(double percent) {
+    return percent * M_PI / 6.0;
```
Comment (asaba96) [inline src/magellan_firmware/firmware/magellan_controller/Drivetrain.cpp]: The formula here and on line 40 aren’t equivalent. How are these derived?

File: src/magellan_firmware/firmware/magellan_controller/EncoderPublisher.cpp [hunk 2]
```diff
@@ -29,26 +29,40 @@ static void SetupEncoderISR() {
...
+            right_delta *= -1;
+        }
 
-    encoder_msg_.header.stamp = nh_.now();
-    encoder_msg_.delta_left = delta_left;
-    encoder_msg_.delta_right = delta_right;
+        // Compute and publish velocity in robot frame
+        double avg_speed = ((left_delta + right_delta) / 2.0) * kDistancePerTick / (1.0 / ENCODER_UPDATE_HZ);
+        twist_msg_.header.stamp = nh_.now();
+        twist_msg_.twist.twist.linear.x = avg_speed;
```
Comment (asaba96) [inline src/magellan_firmware/firmware/magellan_controller/EncoderPublisher.cpp]: Why can’t we use the current steering angle to account for direction?

File: src/magellan_firmware/firmware/magellan_controller/Drivetrain.cpp (hunk 1 above)
Comment (4ndr3w) [reply src/magellan_firmware/firmware/magellan_controller/Drivetrain.cpp]: Yeah, this needs to be cleaned up. I'll update

File: src/magellan_firmware/firmware/magellan_controller/EncoderPublisher.cpp (hunk 2 above)
Comment (asaba96) [reply src/magellan_firmware/firmware/magellan_controller/EncoderPublisher.cpp]: Ignore I realized this wouldn’t work.

File: src/magellan_firmware/firmware/magellan_controller/Drivetrain.cpp (hunk 1 above)
Comment (4ndr3w) [reply src/magellan_firmware/firmware/magellan_controller/Drivetrain.cpp]: To follow up, one was using radians and one was using degrees. Both are in radians now. The number is just from the max angle of the inside wheel in a turn

**Pull request merged**